### PR TITLE
feat(expect): add `expect.addEqualityTesters` api.

### PR DESCRIPTION
**Pull request by eryue0220**
This pr is one of prs to solve #3964, which is to support `expect.addEqualityTester` api.

### CODE REVIEW

File: expect/_custom_equality_tester.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
+// Copyright 2018-2024 the Deno authors. All rights reserved. MIT license.
+
+import { Tester } from "./_types.ts";
+
+const customEqualityTesters: Array<Tester> = [];
+
+export function addCustomEqualityTester(newTesters: Array<Tester>) {
```
Comment (babiabeo) [inline expect/_custom_equality_tester.ts:7]: I think we can use `Tester[]` instead of `Array<Tester>`.

File: expect/expect.ts [hunk 2]
```diff
@@ -212,22 +167,4 @@ export function expect(value: unknown, customMessage?: string): Expected {
...
-};
-
-function isPromiseLike(value: unknown): value is PromiseLike<unknown> {
-  if (value == null) {
-    return false;
-  } else {
-    return typeof ((value as Record<string, unknown>).then) === "function";
-  }
-}
+expect.addEqualityTester = addCustomEqualityTester;
```
Comment (kt3k) [inline expect/expect.ts:170]: Looks like this needs to be `addEqualityTesters` (plural `s`) https://jestjs.io/docs/expect#expectaddequalitytesterstesters

File: expect/_custom_equality_tester_test.ts [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+Deno.test("Basic custom equality test", () => {
+  expect(duration1).toEqual(duration1);
+  expect(duration1).not.toEqual(duration3);
+  expect(duration1).not.toEqual(duration2);
+  expect(duration2).not.toBe(duration1);
+  expect(toIterator([duration1, duration2])).toEqual(
+    toIterator([duration2, duration1]),
+  );
```
Comment (kt3k) [inline expect/_custom_equality_tester_test.ts:69]: If I tested this on jest, it threw with the below error:
```
TypeError: 'caller', 'callee', and 'arguments' properties may not be accessed on strict mode functions or the arguments objects for calls to them
```
What does this case try to test? Also the inverted order of `duration1` and `duration2` intentional?

File: expect/expect.ts (hunk 2 above)
Comment (eryue0220) [reply expect/expect.ts]: Updated.

File: expect/_custom_equality_tester_test.ts (hunk 3 above)
Comment (eryue0220) [reply expect/_custom_equality_tester_test.ts]: Oooh, sorry I carelessly ignored this case. 

This case should _**not**_ equal **actually**.

What the test wants to check includes:
* check yield value in array is pass or not (whatever the order);
* check type is pass or not;
* check the instance indeed value is pass or not;

`Jest` implements this through [iterableEquality](https://github.com/jestjs/jest/blob/main/packages/expect-utils/src/utils.ts#L173), and `vitest` reuses this methods. And currently I don't have many ideas about this, so I marked it `todo`.

Or we also can reused the `jest` implementations?

WDYT?

Comment (kt3k) [reply expect/_custom_equality_tester_test.ts]: Reusing iterableEquality sounds good to me.

BTW this PR looks enough useful at this point. I'm landing this as is.

PRs for the above addition are welcome

Comment (eryue0220) [reply expect/_custom_equality_tester_test.ts]: PR for this case: https://github.com/denoland/deno_std/pull/4286

**Pull request merged**